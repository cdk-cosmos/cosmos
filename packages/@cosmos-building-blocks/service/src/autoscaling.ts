import { Construct, Lazy, Fn } from '@aws-cdk/core';
import { InstanceType, InstanceClass, InstanceSize, CfnLaunchTemplate } from '@aws-cdk/aws-ec2';
import { CfnInstanceProfile } from '@aws-cdk/aws-iam';
import {
  CfnAutoScalingGroup,
  AutoScalingGroup,
  AutoScalingGroupProps,
  Monitoring,
  BlockDevice,
  EbsDeviceVolumeType,
  BlockDeviceVolume,
} from '@aws-cdk/aws-autoscaling';

/**
 * Properties for AutoScalingGroupLT
 */
export interface AutoScalingGroupLTProps extends AutoScalingGroupProps {
  /**
   * Set this to `true` if you want to use Instance Template instead of Launch Config.
   * You can fine tune your default LaunchTemplate with `instanceType`, `launchTemplateOverrides` and `instancesDistribution`
   * @default - false
   */
  readonly useInstanceTemplate?: boolean;
  /**
   * Type of instance to launch. if used with `useInstanceTemplate = true`, it will be applied to LaunchTemplate.
   * Make sure to check and tweak `launchTemplateOverrides` and `instancesDistribution` as the default values set for those
   * properties will overwrite this InstanceType
   */
  readonly instanceType: InstanceType;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Array of LaunchTemplateOverridesProperty which is a subproperty of LaunchTemplate that describes an override for a launch template.
   * Currently, the only supported override is instance type
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html
   * @default
   * [{ instanceType: "t3.medium" }, { instanceType: "t3.small" }]
   *
   * This will use any Spot instances available, "t3.medium" or "t3.small"
   * based on `spotAllocationStrategy` set in `instancesDistribution`
   */
  readonly launchTemplateOverrides?: Array<CfnAutoScalingGroup.LaunchTemplateOverridesProperty>;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Specifies the distribution of On-Demand Instances and Spot Instances,
   * the maximum price to pay for Spot Instances, and how the Auto Scaling group allocates instance types
   * to fulfill On-Demand and Spot capacity.
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-instancesdistribution.html
   * @default - Uses all Spot instances of types specified in property `launchTemplateOverrides`
   *  {
   *    onDemandAllocationStrategy: "prioritized",
   *    onDemandBaseCapacity: 0,
   *    onDemandPercentageAboveBaseCapacity: 0,
   *    spotAllocationStrategy: "capacity-optimized",
   *    spotMaxPrice: props.spotPrice,
   *  }
   */
  readonly instancesDistribution?: Partial<CfnAutoScalingGroup.InstancesDistributionProperty>;
}

/**
 * Derived from Base class AutoScalingGroup with additional functionalities of using Launch Template (LT).
 *
 * Base AutoScalingGroup class uses Launch Config and have no ability to use more advanced Launch Templates
 * and MixedInstancesPolicies which gives ability to use Spot fleet and combination of spot and on-demand
 *
 * This construct uses default spot fleet template settings with just one property `useInstanceTemplate = true`
 *
 * With `launchTemplateOverrides` and `instancesDistribution`, it gives additional fine tuning.
 */
export class AutoScalingGroupLT extends AutoScalingGroup {
  constructor(scope: Construct, id: string, props: AutoScalingGroupLTProps) {
    super(scope, id, props);

    // Check if useInstanceTemplate is set to true. If not, do nothing and rely on Base class to
    // to create AutoScalingGroup using LaunchConfig
    const useInstanceTemplate = props.useInstanceTemplate !== undefined ? props.useInstanceTemplate === true : false;
    if (useInstanceTemplate == true) {
      // Prepare props required for LaunchTemplate
      const instanceType =
        props.instanceType !== undefined
          ? props.instanceType.toString()
          : InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString();

      const imageConfig = props.machineImage.getImage(this);
      const userDataToken = Lazy.stringValue({
        produce: () => Fn.base64(this.userData.render()),
      });
      const securityGroupIds = this.connections.securityGroups.map(sg => sg.securityGroupId);
      const instanceMonitoringMode =
        props.instanceMonitoring !== undefined ? props.instanceMonitoring === Monitoring.DETAILED : undefined;
      const monitoring = {
        enabled: instanceMonitoringMode == undefined ? true : instanceMonitoringMode,
      };

      const iamProfile = this.node.tryFindChild('InstanceProfile') as CfnInstanceProfile;

      // Create LaunchTemplate

      const lTemplate = new CfnLaunchTemplate(this, 'LaunchTemplate', {
        launchTemplateData: {
          imageId: imageConfig.imageId,
          securityGroupIds,
          userData: userDataToken,
          iamInstanceProfile: { arn: iamProfile.attrArn },
          keyName: props.keyName,
          monitoring,
          instanceType: instanceType,
          blockDeviceMappings:
            props.blockDevices !== undefined ? synthesizeBlockDeviceMappings(this, props.blockDevices) : undefined,
        },
      });
      lTemplate.node.addDependency(this.role);

      // Prepare props for ASG with MixedInstancesPolicy

      const overrides =
        props.launchTemplateOverrides == undefined
          ? [
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM).toString(),
              },
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString(),
              },
            ]
          : props.launchTemplateOverrides;

      const instancesDistribution = {
        onDemandAllocationStrategy: 'prioritized',
        onDemandBaseCapacity: 0,
        onDemandPercentageAboveBaseCapacity: 0,
        spotAllocationStrategy: 'capacity-optimized',
        spotMaxPrice: props.spotPrice,
        ...props.instancesDistribution,
      };

      // Mixed instance Policy
      const mixedInstancesPolicy: CfnAutoScalingGroup.MixedInstancesPolicyProperty = {
        launchTemplate: {
          launchTemplateSpecification: {
            version: lTemplate.attrLatestVersionNumber,
            launchTemplateId: lTemplate.ref,
          },
          overrides,
        },
        instancesDistribution,
      };

      // Remove LaunchConfig from the resulting CloudFormation template
      if (!this.node.tryRemoveChild('LaunchConfig')) throw new Error('Launch config not found');
      const asg = this.node.tryFindChild('ASG') as CfnAutoScalingGroup;
      asg.addPropertyDeletionOverride('LaunchConfigurationName');
      // Inject mixedInstancesPolicy to use LaunchTemplate
      asg.mixedInstancesPolicy = mixedInstancesPolicy;
    }
  }
}
function synthesizeBlockDeviceMappings(
  construct: Construct,
  blockDevices: BlockDevice[]
): CfnLaunchTemplate.BlockDeviceMappingProperty[] {
  return blockDevices.map<CfnLaunchTemplate.BlockDeviceMappingProperty>(({ deviceName, volume, mappingEnabled }) => {
    const { virtualName, ebsDevice: ebs } = volume;

    if (volume === BlockDeviceVolume.noDevice() || mappingEnabled === false) {
      return {
        deviceName,
        noDevice: 'true',
      };
    }

    if (ebs) {
      const { iops, volumeType } = ebs;

      if (!iops) {
        if (volumeType === EbsDeviceVolumeType.IO1) {
          throw new Error('iops property is required with volumeType: EbsDeviceVolumeType.IO1');
        }
      } else if (volumeType !== EbsDeviceVolumeType.IO1) {
        construct.node.addWarning('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
      }
    }

    return {
      deviceName,
      ebs,
      virtualName,
    };
  });
}
