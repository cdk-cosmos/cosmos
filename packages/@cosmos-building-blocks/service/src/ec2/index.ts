export {
  CloudWatchAgent,
  CloudWatchAgentProps,
  CloudWatchMetricBuilder,
  CloudWatchMetric,
} from './ec2-cloud-watch-agent';
