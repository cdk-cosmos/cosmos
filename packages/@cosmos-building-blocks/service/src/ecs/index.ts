export { EcsService, EcsServiceProps } from './ecs-service';
