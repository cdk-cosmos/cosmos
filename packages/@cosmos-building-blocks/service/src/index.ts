export { EcsService, EcsServiceProps } from './ecs';
