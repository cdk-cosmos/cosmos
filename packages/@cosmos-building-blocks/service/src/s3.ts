import { Construct } from '@aws-cdk/core';
import { Bucket, BucketEncryption, BucketProps, BucketAccessControl } from '@aws-cdk/aws-s3';
import { PolicyStatement, Effect, AnyPrincipal } from '@aws-cdk/aws-iam';

export class SecureBucket extends Bucket {
  constructor(scope: Construct, id: string, props?: BucketProps) {
    super(scope, id, {
      encryption: BucketEncryption.S3_MANAGED,
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      serverAccessLogsPrefix: 'BucketAccessLogs/',
      ...props,
    });
    const policyStatement = new PolicyStatement({
      actions: ['s3:*'],
      principals: [new AnyPrincipal()],
      resources: [`${this.bucketArn}/*`],
      effect: Effect.DENY,
      sid: 'ForceSSLOnly',
    });
    policyStatement.addCondition('Bool', { 'aws:SecureTransport': false });
    super.addToResourcePolicy(policyStatement);
  }
}
