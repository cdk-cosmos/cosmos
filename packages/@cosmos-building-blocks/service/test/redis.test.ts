import { SynthUtils, expect as cdkExpect, haveResourceLike } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Redis } from '../src/index';
import { Vpc, SubnetType, Instance, InstanceClass, InstanceType, InstanceSize, MachineImage } from '@aws-cdk/aws-ec2';
import '@aws-cdk/assert/jest';

test('Should match snapshot', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'aws-redis');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
      {
        name: 'Redis',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 28,
      },
    ],
  });
  const ec2 = new Instance(stack, 'Ec2', {
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.LARGE),
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
  });
  // Redis cluster only allowing connections from ec2 in 'Redis' subnet
  new Redis(stack, 'Redis', {
    vpc,
    connectionsAllowedFrom: ec2.connections,
    subnetIds: [{ subnetGroupName: 'Redis' }],
  });
  // Redis cluster allowing all connections in 'Redis' subnet
  new Redis(stack, 'RedisOpen', {
    vpc,
    subnetIds: [{ subnetGroupName: 'Redis' }],
  });

  // Redis cluster overwriting some of the default props
  new Redis(stack, 'RedisOverwrite', {
    vpc,
    subnetIds: [{ subnetGroupName: 'App' }],
    replicationGroupDescription: 'Redis cluster not encrypted',
    atRestEncryptionEnabled: false,
    cacheNodeType: 'cache.t3.medium',
    numCacheClusters: 3,
    transitEncryptionEnabled: false,
    automaticFailoverEnabled: true,
  });
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Should be able to overwrite props', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'aws-redis');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
      {
        name: 'Redis',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 28,
      },
    ],
  });
  new Redis(stack, 'RedisTest2', {
    vpc,
    subnetIds: [{ subnetGroupName: 'App' }],
    atRestEncryptionEnabled: true,
    cacheNodeType: 'cache.t3.medium',
    numCacheClusters: 3,
    transitEncryptionEnabled: false,
    automaticFailoverEnabled: true,
    engine: 'memcached', // This should not create memcache engine and should overwrite to Redis
  });
  // THEN
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { AtRestEncryptionEnabled: true });
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { TransitEncryptionEnabled: false });
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { NumCacheClusters: 3 });
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { CacheNodeType: 'cache.t3.medium' });
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { AutomaticFailoverEnabled: true });
  expect(stack).toHaveResource('AWS::ElastiCache::ReplicationGroup', { Engine: 'redis' }); // Engine should always be redis
});

test('Default Redis should have all default properties', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'redis-default');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
      {
        name: 'Redis',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 28,
      },
    ],
  });
  // Redis with only required props
  new Redis(stack, 'RedisTest3', {
    vpc,
    subnetIds: [{ subnetGroupName: 'Redis' }],
  });
  // THEN
  // Redis SG should allow connection from anywhere on port 6379
  cdkExpect(stack).to(
    haveResourceLike('AWS::EC2::SecurityGroup', {
      GroupDescription: 'redis-default/RedisTest3/RedisSecurityGroup',
      SecurityGroupIngress: [{ CidrIp: '0.0.0.0/0', ToPort: 6379 }],
    })
  );
  // Redis SubnetGroup should be created only in subnets provided in props
  cdkExpect(stack).to(
    haveResourceLike('AWS::ElastiCache::SubnetGroup', {
      SubnetIds: [
        {
          Ref: 'vpcRedisSubnet1Subnet3EE087E6',
        },
        {
          Ref: 'vpcRedisSubnet2SubnetFF7CC220',
        },
      ],
    })
  );
});
