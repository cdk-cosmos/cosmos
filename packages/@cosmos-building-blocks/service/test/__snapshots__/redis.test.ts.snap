// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "Ec2FD86B82F": Object {
      "DependsOn": Array [
        "Ec2InstanceRole8071A9F0",
      ],
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "IamInstanceProfile": Object {
          "Ref": "Ec2InstanceProfile7C946CD9",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.large",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "Ec2InstanceSecurityGroup1638B7F2",
              "GroupId",
            ],
          },
        ],
        "SubnetId": Object {
          "Ref": "vpcAppSubnet1Subnet0009B9D9",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/Ec2",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "Ec2InstanceProfile7C946CD9": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "Ec2InstanceRole8071A9F0",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "Ec2InstanceRole8071A9F0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/Ec2",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2InstanceSecurityGroup1638B7F2": Object {
      "Properties": Object {
        "GroupDescription": "aws-redis/Ec2/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/Ec2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "Redis04B3B4F3": Object {
      "Properties": Object {
        "AtRestEncryptionEnabled": true,
        "AutomaticFailoverEnabled": false,
        "CacheNodeType": "cache.t3.micro",
        "CacheSubnetGroupName": Object {
          "Ref": "RedisRedisSubnetGroupE7D796E2",
        },
        "Engine": "redis",
        "MultiAZEnabled": false,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 0,
        "ReplicationGroupDescription": "Redis cluster for aws-redis/Redis",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "RedisRedisSecurityGroup23991694",
              "GroupId",
            ],
          },
        ],
        "TransitEncryptionEnabled": true,
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "RedisClosedRedis1D8745B1": Object {
      "Properties": Object {
        "AtRestEncryptionEnabled": true,
        "AutomaticFailoverEnabled": false,
        "CacheNodeType": "cache.t3.micro",
        "CacheSubnetGroupName": Object {
          "Ref": "RedisClosedRedisSubnetGroupD85ADBD5",
        },
        "Engine": "redis",
        "MultiAZEnabled": false,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 0,
        "ReplicationGroupDescription": "Redis cluster for aws-redis/RedisClosed",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "RedisClosedRedisSecurityGroup49D34B18",
              "GroupId",
            ],
          },
        ],
        "TransitEncryptionEnabled": true,
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "RedisClosedRedisSecurityGroup49D34B18": Object {
      "Properties": Object {
        "GroupDescription": "Redis Security Group for aws-redis/RedisClosed",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RedisClosedRedisSecurityGroupfromawsredisEc2InstanceSecurityGroup964FF110637921FDFAF6": Object {
      "Properties": Object {
        "Description": "from awsredisEc2InstanceSecurityGroup964FF110:6379",
        "FromPort": 6379,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "RedisClosedRedisSecurityGroup49D34B18",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "Ec2InstanceSecurityGroup1638B7F2",
            "GroupId",
          ],
        },
        "ToPort": 6379,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "RedisClosedRedisSubnetGroupD85ADBD5": Object {
      "Properties": Object {
        "Description": "Redis Subnet group for aws-redis/RedisClosed",
        "SubnetIds": Array [
          Object {
            "Ref": "vpcRedisSubnet1Subnet3EE087E6",
          },
          Object {
            "Ref": "vpcRedisSubnet2SubnetFF7CC220",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "RedisHaRedisE071ED75": Object {
      "Properties": Object {
        "AtRestEncryptionEnabled": true,
        "AutomaticFailoverEnabled": true,
        "CacheNodeType": "cache.t3.micro",
        "CacheSubnetGroupName": Object {
          "Ref": "RedisHaRedisSubnetGroupBB091AA5",
        },
        "Engine": "redis",
        "MultiAZEnabled": true,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 1,
        "ReplicationGroupDescription": "Redis cluster for aws-redis/RedisHa",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "RedisHaRedisSecurityGroup8540882A",
              "GroupId",
            ],
          },
        ],
        "TransitEncryptionEnabled": true,
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "RedisHaRedisSecurityGroup8540882A": Object {
      "Properties": Object {
        "GroupDescription": "Redis Security Group for aws-redis/RedisHa",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:6379",
            "FromPort": 6379,
            "IpProtocol": "tcp",
            "ToPort": 6379,
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RedisHaRedisSubnetGroupBB091AA5": Object {
      "Properties": Object {
        "Description": "Redis Subnet group for aws-redis/RedisHa",
        "SubnetIds": Array [
          Object {
            "Ref": "vpcRedisSubnet1Subnet3EE087E6",
          },
          Object {
            "Ref": "vpcRedisSubnet2SubnetFF7CC220",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "RedisOverwriteRedis9B423C14": Object {
      "Properties": Object {
        "AtRestEncryptionEnabled": false,
        "AutomaticFailoverEnabled": true,
        "CacheNodeType": "cache.t3.medium",
        "CacheSubnetGroupName": Object {
          "Ref": "RedisOverwriteRedisSubnetGroupB4851474",
        },
        "Engine": "redis",
        "MultiAZEnabled": false,
        "NumCacheClusters": 3,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 0,
        "ReplicationGroupDescription": "Redis cluster not encrypted",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "RedisOverwriteRedisSecurityGroupDBCF27F8",
              "GroupId",
            ],
          },
        ],
        "TransitEncryptionEnabled": false,
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "RedisOverwriteRedisSecurityGroupDBCF27F8": Object {
      "Properties": Object {
        "GroupDescription": "Redis Security Group for aws-redis/RedisOverwrite",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:6379",
            "FromPort": 6379,
            "IpProtocol": "tcp",
            "ToPort": 6379,
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RedisOverwriteRedisSubnetGroupB4851474": Object {
      "Properties": Object {
        "Description": "Redis Subnet group for aws-redis/RedisOverwrite",
        "SubnetIds": Array [
          Object {
            "Ref": "vpcAppSubnet1Subnet0009B9D9",
          },
          Object {
            "Ref": "vpcAppSubnet2Subnet01033D46",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "RedisRedisSecurityGroup23991694": Object {
      "Properties": Object {
        "GroupDescription": "Redis Security Group for aws-redis/Redis",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:6379",
            "FromPort": 6379,
            "IpProtocol": "tcp",
            "ToPort": 6379,
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RedisRedisSubnetGroupE7D796E2": Object {
      "Properties": Object {
        "Description": "Redis Subnet group for aws-redis/Redis",
        "SubnetIds": Array [
          Object {
            "Ref": "vpcRedisSubnet1Subnet3EE087E6",
          },
          Object {
            "Ref": "vpcRedisSubnet2SubnetFF7CC220",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "vpcA2121C38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcAppSubnet1RouteTable883411EE": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/AppSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcAppSubnet1RouteTableAssociation7C392A32": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcAppSubnet1RouteTable883411EE",
        },
        "SubnetId": Object {
          "Ref": "vpcAppSubnet1Subnet0009B9D9",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcAppSubnet1Subnet0009B9D9": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "App",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/AppSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcAppSubnet2RouteTable9E7B4A01": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/AppSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcAppSubnet2RouteTableAssociationD024F74B": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcAppSubnet2RouteTable9E7B4A01",
        },
        "SubnetId": Object {
          "Ref": "vpcAppSubnet2Subnet01033D46",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcAppSubnet2Subnet01033D46": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.64/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "App",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/AppSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcRedisSubnet1RouteTable9EFB9850": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/RedisSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcRedisSubnet1RouteTableAssociation2DCFFBEC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcRedisSubnet1RouteTable9EFB9850",
        },
        "SubnetId": Object {
          "Ref": "vpcRedisSubnet1Subnet3EE087E6",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcRedisSubnet1Subnet3EE087E6": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.128/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Redis",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/RedisSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcRedisSubnet2RouteTable262EC3FD": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/RedisSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcRedisSubnet2RouteTableAssociationAFAC6951": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcRedisSubnet2RouteTable262EC3FD",
        },
        "SubnetId": Object {
          "Ref": "vpcRedisSubnet2SubnetFF7CC220",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcRedisSubnet2SubnetFF7CC220": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.144/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Redis",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-redis/vpc/RedisSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
